### PR TITLE
+ Storybook & AssetLoader

### LINKED ISSUE #38: Resources url shouldn't be hardcoded
Resources url is hardcoded to localhost, leaving no room to actually use the renderer. The url should be customizable:
https://github.com/kozennnn/scuti-renderer/blob/dev/src/Scuti.ts#L92

```js
        Assets.add("room/materials", "http://127.0.0.1:8081/generic/room/room_data.json");
        Assets.add("room/room", "http://127.0.0.1:8081/generic/room/room.json");
        Assets.add("room/cursors", "http://127.0.0.1:8081/generic/tile_cursor/tile_cursor.json");
        Assets.add("furnitures/floor/placeholder", "http://127.0.0.1:8081/generic/place_holder/place_holder_furniture.json");
        Assets.add("furnitures/furnidata", "http://127.0.0.1:8081/gamedata/furnidata.json");
        Assets.add("figures/figuredata", "http://127.0.0.1:8081/gamedata/figuredata.json");
        Assets.add("figures/figuremap", "http://127.0.0.1:8081/gamedata/figuremap.json");
        Assets.add("figures/draworder", "http://127.0.0.1:8081/gamedata/draworder.json");
        Assets.add("figures/actions", "http://127.0.0.1:8081/generic/HabboAvatarActions.json");
        Assets.add("figures/partsets", "http://127.0.0.1:8081/generic/HabboAvatarPartSets.json");
        Assets.add("figures/animations", "http://127.0.0.1:8081/generic/HabboAvatarAnimations.json");
```

I've used GitHub pages to host the resources at https://psociety.github.io/scuti-resources/ it that helps. At least it would be more accessible than expecting localhost self serving. It should also be done at https://github.com/kozennnn/scuti-resources so it can officially be used.

### PR DESCRIPTION
**Pull request by tetreum**
This PR:
- Adds storybook: It will be useful to list usage examples & debugging scenes
- Fixes build process: Until now the renderer wasn't compiled as a package.
- Adds AssetLoader, that facilitates loading cdn assets from any class.

image

Closes https://github.com/kozennnn/scuti-renderer/issues/38 & https://github.com/kozennnn/scuti-renderer/issues/37

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,19 +1,26 @@
 {
-  "name": "scuti-renderer",
+  "name": "@kozennnn/scuti-renderer",
```
Comment (tetreum) [inline package.json:2]: needs to have the same name as the npm so it can be locally required too.

File: public/index.html [hunk 2]
```diff
@@ -11,6 +11,16 @@
...
+        div#stats {
+            position: fixed;
+            top: 0;
+            right: 0;
+            z-index: 500;
+            width: max(200px, 10vw, 10vh);
+            height: max(100px, 6vh, 6vw);
+            opacity: 0.8;
+            user-select: none;
+        }
```
Comment (tetreum) [inline public/index.html:23]: i need this to make my life easier when debugging, so i don't have to manually add it everytime. (from https://www.npmjs.com/package/pixi-stats)

File: src/objects/rooms/RoomTileMap.ts [hunk 3]
```diff
@@ -39,6 +39,8 @@ export class RoomTileMap {
     private _parse(
         tileMap: string
     ): TileMap {
+        tileMap = tileMap.replace(/ /g, '');
+        tileMap = tileMap.replace(/\n\n/g, '\n');
```
Comment (tetreum) [inline src/objects/rooms/RoomTileMap.ts:43]: this way we can pass single string tileMaps like:
```js
tileMap: `
    xxxxx
    x1100
    x1100
    x0000
    x0000
    `
```

File: storybook/src/stories/Room.stories.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    "x0000000000000000x\n" +
+    "00000000000000000x\n" +
+    "x0000000000000000x\n" +
+    "x0000000000000000x\n" +
+    "x0000000000000000x\n" +
+    "x0000000000000000x\n" +
+    "x0000000000000000x\n" +
+    "x0000000000000000x\n",
+  },
+};
```
Comment (tetreum) [inline storybook/src/stories/Room.stories.ts]: here, or on to a whole dedicated stories page,  you can add any tileMap tests that you had like in script2.js

File: storybook/src/stories/Room.vue [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+  
+        if (container == null) return;
+  
+        scuti = new Scuti({
+          canvas: container,
+          width: window.innerWidth / 2,
+          height: window.innerHeight / 2,
+          resources: './resources'
+        });
+        await scuti.loadResources("https://psociety.github.io/scuti-resources/");
```
Comment (tetreum) [inline storybook/src/stories/Room.vue]: I'm pointing to psociety's repo but the official `scuti-resources` should have GitHub Pages activated so we can point to it.

File: storybook/src/stories/Room.vue [hunk 6]
```diff
@@ -0,0 +1,48 @@
+<template>
```
Comment (tetreum) [inline storybook/src/stories/Room.vue]: Here is where i failed to achieve what i wanted. I wanted a Vue Component that RETURNs `scuti` var, so there's more room for customization on each story. But i don't have experience with Vue.

I would have hacked it by simply adding a callback as prop that returns scuti var, but i don't know if there's a better way to achieve this.

File: src/Scuti.ts [hunk 7]
```diff
@@ -87,31 +88,23 @@ export class Scuti {
...
-        await Assets.load("room/room");
-        await Assets.load("room/cursors");
-        await Assets.load("furnitures/floor/placeholder");
-        await Assets.load("furnitures/furnidata");
-        await Assets.load("figures/figuredata");
-        await Assets.load("figures/figuremap");
-        await Assets.load("figures/draworder");
-        await Assets.load("figures/actions");
-        await Assets.load("figures/partsets");
-        await Assets.load("figures/animations");
```
Comment (tetreum) [inline src/Scuti.ts:114]: if you call `awaits`  one after another you're sequentially loading them. With Promise.All, all start loading at the same time while it still waits for all of them finish before continuing.
So it should be faster now.

File: storybook/src/stories/Renderer.vue [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  
+        if (container == null) return;
+  
+        scuti = new Scuti({
+          canvas: container,
+          width: window.innerWidth / 2,
+          height: window.innerHeight / 2,
+          resources: './resources'
+        });
+        await scuti.loadResources("https://psociety.github.io/scuti-resources/");
```
Comment (tetreum) [inline storybook/src/stories/Renderer.vue:28]: I'm pointing to psociety's repo but the official `scuti-resources``should have GitHub Pages activated so we can point to it.

File: storybook/src/stories/Room.stories.ts [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+            x0000
+            x0000
+        `,
+          floorMaterial: new FloorMaterial(scuti, 110),
+          wallMaterial: new WallMaterial(scuti, 1601)
+      });
+    },
+  },
+};
+
```
Comment (tetreum) [inline storybook/src/stories/Room.stories.ts:51]: here, or on to a whole dedicated story, page you can add any tileMap tests that you had like in script2.js

File: storybook/src/stories/Renderer.vue (hunk 8 above)
Comment (mathishouis) [reply storybook/src/stories/Renderer.vue:28]: I'm going to do that this evening 😸

Comment (tetreum) [reply storybook/src/stories/Renderer.vue:28]: will be easy, it's just clicking a button, as you just have to select which branch will be published as a web :)

image

Comment (mathishouis) [reply storybook/src/stories/Renderer.vue:28]: Thanks, it is available at this URL: [https://kozennnn.github.io/scuti-resources/](https://kozennnn.github.io/scuti-resources/) :D

**Pull request merged**